msprime/tables.py

```diff
@@ -53,33 +54,86 @@ def text_decode_metadata(encoded):
+    "SiteTableRow",
+    ["position", "ancestral_state", "metadata"])
+
+
+MutationTableRow = collections.namedtuple(
+    "MutationTableRow",
+    ["site", "node", "derived_state", "parent", "metadata"])
+
+
+ProvenanceTableRow = collections.namedtuple(
+    "ProvenanceTableRow",
+    ["timestamp", "record"])
+
+
 class NodeTable(_msprime.NodeTable):
     """
-    Class for tables describing all nodes in the tree sequence, of the form
-        id     is_sample  population   time
-        0      1          0            0.0
-        1      1          1            0.0
-        2      0          0            0.0
-        3      1          0            0.5
-        4      0          2            2.1
-    Node IDs are *not* recorded; rather the `id` column shows the row index, so
-    that the `k`-th row describes the node whose ID is `k`.  `is_sample`
-    records whether the node is a sample (=1) or not (=0).  `population` is an
-    integer population ID, and `time` is the time since that individual was
-    born, as a float.
-
-    Requirements:
-        1. All birth times must be greater than or equal to zero.
-
-    It is not required that the `time` column be ordered or that all samples
-    must be at the top.
+    A table defining the nodes in a tree sequence. See the
+    :ref:`definitions <sec-node-table-definition>` for details on the columns
+    in this table and the
+    :ref:`tree sequence requirements <sec-valid-tree-sequence-requirements>` section
+    for the properties needed for a node table to be a part of a valid tree sequence.
+
+    Example::
+
+        >>> t = msprime.NodeTable()
+        >>> t.add_row(flags=0, time=1)
+        0
+        >>> t.add_row(flags=1, time=2)
+        1
+        >>> print(t)
+        id      flags   population      time    metadata
+        0       0       -1      1.00000000000000
+        1       1       -1      2.00000000000000
+        >>> print(t[0])
+        NodeTableRow(flags=0, time=1.0, population=-1, metadata=b'')
+        >>> print(t[-1])
+        NodeTableRow(flags=1, time=2.0, population=-1, metadata=b'')
+        >>> len(t)
+        2
+
+    :warning: The numpy arrays returned by table attribute accesses are **copies**
+        of the underlying data. In particular, this means that you cannot edit
+        the values in the columns by updating the attribute arrays.
+
+        **NOTE:** this behaviour may change in future.
+
+    :ivar time: The array of time values.
+    :vartype time: numpy.ndarray, dtype=np.float64
+    :ivar flags: The array of flags values.
+    :vartype flags: numpy.ndarray, dtype=np.uint32
+    :ivar population: The array of population IDs.
+    :vartype population: numpy.ndarray, dtype=np.int32
+    :ivar metadata: The flattened array of binary metadata values.
+    :vartype metadata: numpy.ndarray, dtype=np.int8
+    :ivar metadata_offset: The array of offsets into the metadata column.
+    :vartype metadata_offset: numpy.ndarray, dtype=np.uint32
```

Comment:
should http://msprime.readthedocs.io/en/docathon/interchange.html#sec-encoding-ragged-columns be cross-referenced in the class-level docstring too?

Reply:
Good idea. I'll see what I can do.